spinn_front_end_common/interface/interface_functions/insert_edges_to_extra_monitor_functionality.py

```diff
@@ -0,0 +1,135 @@
+from pacman.model.graphs.application import ApplicationEdge
+from pacman.model.graphs.machine import MachineEdge
+from spinn_front_end_common.utilities import constants
+from spinn_front_end_common.utility_models.\
+    data_speed_up_packet_gatherer_machine_vertex import \
+    DataSpeedUpPacketGatherMachineVertex
+from spinn_front_end_common.utility_models.\
+    extra_monitor_support_application_vertex import \
+    ExtraMonitorSupportApplicationVertex
+from spinn_front_end_common.utility_models.\
+    extra_monitor_support_machine_vertex import \
+    ExtraMonitorSupportMachineVertex
+from spinn_utilities.progress_bar import ProgressBar
+
+
+class InsertEdgesToExtraMonitorFunctionality(object):
+
+    def __call__(self, machine_graph, placements, machine,
+                 vertex_to_ethernet_connected_chip_mapping,
+                 application_graph=None, graph_mapper=None):
+        """ inserts edges between verts whom use mc speed up and its local\
+         mc data gatherer
+
+        :param machine_graph: the machine graph instance
+        :param placements: the placements
+        :param machine: the machine object
+        :param application_graph: the application graph
+        :param vertex_to_ethernet_connected_chip_mapping: mapping between \
+        ethernet connected chips and packet gatherers
+        :param graph_mapper: the graph mapper
+        :rtype: None
+        """
+        n_app_vertices = 0
+        if application_graph is not None:
+            n_app_vertices = len(application_graph.vertices)
+
+        progress = ProgressBar(
+            len(machine_graph.vertices) + n_app_vertices,
+            "Inserting edges between vertices which require fr speed up "
+            "functionality. ")
+
+        for vertex in progress.over(machine_graph.vertices, False):
+            if isinstance(vertex, ExtraMonitorSupportMachineVertex):
+                self._process_vertex(
+                    vertex, machine, placements, machine_graph,
+                    vertex_to_ethernet_connected_chip_mapping,
+                    application_graph, graph_mapper)
+
+        if application_graph is not None:
+            for vertex in progress.over(application_graph.vertices):
+                if isinstance(vertex, ExtraMonitorSupportApplicationVertex):
+                    machine_verts = graph_mapper.get_machine_vertices(vertex)
+                    for machine_vertex in machine_verts:
+                        self._process_vertex(
+                            machine_vertex, machine, placements, machine_graph,
+                            vertex_to_ethernet_connected_chip_mapping,
+                            application_graph, graph_mapper)
+        progress.end()
```

Comment:
not seeing your issue here. 
starts by progressing over machine verts without ending the process bar. 
now if theres no app graph, it needs to end the progress thus line 58. But if theres a app graph, then then iterate over it. end afterwards to ensure no dangling things. 

you could maybe argue putting a flase in the progress.over(app verts) but i dont think calling progressbar end multiple times is that big aof a issue. but we can if that makes ya better

Reply:
It was to do with progress bar API usage; there was an `over()` call just above that needed a `False` second parameter passing in so that the termination of iteration wouldn't trigger a call to `end()` (which is only supposed to be called once).

Reply:
ok, fixed